typegen/src/lib.rs

```diff
@@ -0,0 +1,29 @@
+// Copyright 2019-2022 Parity Technologies (UK) Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+pub mod typegen;
+pub mod utils;
+
+pub use typegen::{
+    error::TypegenError,
+    settings::{
+        derives::{Derives, DerivesRegistry},
+        substitutes::TypeSubstitutes,
+        TypeGeneratorSettings,
+    },
+    TypeGenerator,
+};
```

Comment:
A general Q: there is still quite a lot of random stuff exposed, and I wonder whether we can avoid exposing some of it to make the surface area of this lib as small as possible (the ideal place to get to would be that we are happy with the interface on everything that is exposed; are we?)

Reply:
I think we could cut out the `Derives` from public access. But the `TypeGeneratorSettings` should be exposed and I would like to have all of its fields `pub` as well, to create custom settings easily without relying on some builder functions. `DerivesRegistry` and `TypeSubstitutes` are fields of the settings, so we would need to have them public as well. Maybe the names are not so nice to expose like this though. Maybe something like `DeriveSettings` and `SubstituteSettings`?